Docs release 1.8.0 csi 3573 external log collection (PR2) (#418)
IBM/ibm-block-csi-driver: docs/content/troubleshooting/log_status_collect.md

To collect and display status and logs related to the different components of IBM® block storage CSI driver, use the Kubernetes commands that are found in this section.

Before you begin collecting logs, create a directory for the logs on the control-plane or bastion server.

For example:

```
mkdir logs
```
Save logs and status reports directly to the created directory by adding in the following string at the end of the collection command: `> logs/<log_filename>`.

**Important:** Be sure that the logs cover any relevant time frames for the specific issues that you are trying to debug when gathering logs from the storage system.

**Note:** All commands here are listed with the collection command with the `logs` folder name example. Change the folder name according as needed.

### General status and log collection for IBM Support
Be sure to run the following steps and copy the output to an external file, when engaging IBM Support and sending log collections.

1. Check the node status.
    
    `kubectl get nodes`
2. Check the CSI driver component status.

    `kubectl get -all-namespaces pod -o wide | grep ibm-block-csi`
3. Check if the PersistentVolumeClaims (PVCs) are _Bound_.

>     `kubectl get pvc`

Reviewer: btw the `.metadata.name` hides the status (e.g. `Bound`), which is the point of this step.
so @roysahar we might need to change to command a bit

>     `kubectl get pvc`

Reviewer 2: @oriyarde / @rivkap , I will check for showing also "Bound" state.

>     `kubectl get pvc`

Reviewer: thanks

>     `kubectl get -n <pvcs_namespace> pvc -o=jsonpath='{.items[?(@.metadata.annotations.volume\.beta\.kubernetes\.io/storage-provisioner=="block.csi.ibm.com")].metadata.name}'`

Reviewer 2: @rivkap please change the command to:
`kubectl get -n default pvc -o=jsonpath='{range .items[?(@.metadata.annotations.volume\.beta\.kubernetes\.io/storage-provisioner=="block.csi.ibm.com")]}{"PVC NAME: "}{@.metadata.name}{" PVC STATUS: "}{@.status.phase}{"\n"}{end}'`

It will give an example output:
`PVC NAME: demo-pvc-file-system PVC STATUS: Bound`

Reviewer: LGTM

Author: done. thanks @roysahar

    
    for pod in $nodepods;do for container in `kubectl get -n <namespace> pod $pod -o jsonpath='{.spec.containers[*].name}'`;do kubectl logs -n <namespace> $pod -c $container > logs/${pod}_${container}.log;done;done

### Log collection for all CSI controller containers

To collect logs for all pods and containers, use the following commands:
    
    for container in `kubectl get -n <namespace> pod ibm-block-csi-controller-0 -o jsonpath='{.spec.containers[*].name}'`;do kubectl logs -n <namespace> ibm-block-csi-controller-0 -c $container > logs/ibm-block-csi-controller-0_${container}.log;done

### Log collection for CSI operator logs
To collect CSI operator logs, use the following commands:

    operatorpod=`kubectl get pods --all-namespaces |grep ibm-block-csi-operator|awk '{print $2}'`
    kubectl logs $operatorpod -n <namespace> > logs/operator

### Collecting details of all CSI objects and components
`kubectl describe all -l product=ibm-block-csi-driver -n <namespace> > logs/describe_ibm-block-csi-driver`

### Status collection for CSI pods, daemonset, and statefulset

> `kubectl get all -n <namespace>  -l product=ibm-block-csi-driver > logs/get_all_ibm-block-csi-driver`

Reviewer: ```suggestion
`kubectl get all -n <namespace> -l product=ibm-block-csi-driver > logs/get_all_ibm-block-csi-driver`
```

Author: done

> ### Log collection for the CSI driver controller

Reviewer 2: Why do we need this? We have the same and more in line 63.

Author: @roysahar  Does 
`for container in `kubectl get -n <namespace> pod ibm-block-csi-controller-0 -o jsonpath='{.spec.containers[*].name}'`;do kubectl logs -n <namespace> ibm-block-csi-controller-0 -c $container > logs/ibm-block-csi-controller-0_${container}.log;done`

do the same thing as 
`kubectl logs -f -n <namespace> ibm-block-csi-controller-0 -c ibm-block-csi-controller > logs/ibm-block-csi-controller``

Reviewer: the `for` loop makes the output per file, which is important when dealing with support

Author: As per my discussion with @oriyarde  I am going to indent three of the "all" commands under the commands for Support.

The other "smaller" similar commands are for the user to use so that they have more directed log output to troubleshoot specific needs.

Reviewer 2: But the "small" command will show only the one of the containers (`ibm-block-csi-controller`) in the `ibm-block-csi-controller-0` POD. This can lead the customer not to see the reason for the failure/s he is looking.

Author: I think that the information is there and they can always run the other commands but this way they're not overwhelmed with information.  the support commands that we're giving are almost a "courtesy" to support so that they get the fullest picture possible when contacted.

@oriyarde  what do you think? leave or remove?

Reviewer 2: OK

`kubectl logs -f -n <namespace> ibm-block-csi-controller-0 -c ibm-block-csi-controller > logs/ibm-block-csi-controller`

> ### Log collection for the CSI driver node (per worker node or PODID)

Reviewer 2: Why do we need this? We have the same and more (for node pods) in line 54.

Author: @roysahar  is what we have for section on line 54

    nodepods=`kubectl get pods -l product=ibm-block-csi-driver -l app.kubernetes.io/component=csi-node --output=jsonpath={.items..metadata.name}`
    
    for pod in $nodepods;do for container in `kubectl get -n <namespace> pod $pod -o jsonpath='{.spec.containers[*].name}'`;do kubectl logs -n <namespace> $pod -c $container > logs/${pod}_${container}.log;done;done

better than

`kubectl logs -f -n <namespace> ibm-block-csi-node-<PODID> -c ibm-block-csi-node > logs/csi-node-<PODID>`

Author: As per my discussion with @oriyarde  I am going to indent three of the "all" commands under the commands for Support.

The other "smaller" similar commands are for the user to use so that they have more directed log output to troubleshoot specific needs.

Author: commented above as well

Reviewer 2: OK

`kubectl logs -f -n <namespace> ibm-block-csi-node-<PODID> -c ibm-block-csi-node > logs/csi-node-<PODID>`

### Details collection for unbound PVCs

> `kubectl describe pvc <pvc-name> > logs/pvc_not_bounded`

Reviewer 2: Maybe we need to add `-n <pvcs_namespace>`

Reviewer 2: `kubectl describe -n <pvc_namespace> pvc <pvc-name> > logs/pvc_not_bounded`

Author: done

### Details collection for pods not in the _Running_ state

> `kubectl describe pod <not-running-pod-name> > logs/pod_not_running`

Reviewer 2: Also here, consider to add `-n <pod_namespace>`

Author: Please write the whole command with the -n <pvcs_namespace> so I see exactly where to place it in the command. thanks
@roysahar

Author: done

# Detecting errors and log collection

Use the CSI (Container Storage Interface) driver debug information for problem identification.

**Note:** These procedures are applicable for both Kubernetes and Red Hat® OpenShift®. For Red Hat OpenShift, replace `kubectl` with `oc` in all relevant commands.

## Detecting errors

To help pinpoint potential causes for stateful pod failure:

1.  Verify that all CSI pods are running.
    ```
    kubectl get pods -n <namespace> -l product=ibm-block-csi-driver
    ```

2.  If a pod is not in a _Running_ state, run the following command:
    ```
    kubectl describe -n <namespace> pod/<pod-name>
    ```
    View the logs.

## Status and log collection
To collect and display status and logs related to the different components of IBM® block storage CSI driver, use the Kubernetes commands that are found in this section.

Before you begin collecting logs, create a directory for the logs on the control-plane or bastion server.

For example:

```
mkdir logs
```

> Save logs and status reports directly to the created directory by adding in the following string at the end of collection command: `> logs/<log_filename>`.

Reviewer: ```suggestion
Save logs and status reports directly to the created directory by adding in the following string at the end of the collection command: `> logs/<log_filename>`.
```

Author: DONE

**Important:** Be sure that the logs cover any relevant timeframes for the specific issues that you are trying to debug when gathering logs from the storage system.

### General status and log collection for IBM Support
Be sure to run the following steps and copy the output to an external file, when engaging IBM Support and sending log collections.

1. Check the node status.
    
    `kubectl get nodes`
2. Check the CSI driver component status.

    `kubectl get -n <namespace> pod -o wide | grep ibm-block-csi`
3. Check if the PersistentVolumeClaims (PVCs) are _Bound_.

    `kubectl get pvc`

    - If the PVCs are not in the _Bound_ state collect the events of all unbound PVCs. (See [Log collection for unbound PVCs](#log-collection-for-unbound-pvcs).)

### Log collection for all CSI nodepods and their containers

> To collect logs for all nodepods, use the following commands:

Reviewer: ```suggestion
To collect logs for all node pods, use the following commands:
```
I don't think `nodepods` is a term

Author: DONE

>     nodepods=`kubectl get -n <namespace> pod -l app.kubernetes.io/component=csi-node --output=jsonpath={.items..metadata.name}`

Reviewer: @roysahar is this label unique to ibm block csi driver?

Author: Fixed `kubectl get pods -l product=ibm-block-csi-driver -l app.kubernetes.io/component=csi-node --output=jsonpath={.items..metadata.name}`
